### ISSUE TITLE
Extract deepEqual implementation with tests to share it with other libraries?

### ISSUE THREAD

**Issue by jzaefferer** [qunitjs/qunit#865]
At least node has its own implementation of deepEqual, as part of the assert module. I have no idea how close their implementation is to ours, might be worth investigating.

**Comment by jzaefferer** [comment]
Looking around a bit, I found three distinct implementations, all with their own tests. There's definitely potential to share at least tests, better yet implementation.

- substack's node-deep-equal [implementation](https://github.com/substack/node-deep-equal/blob/master/index.js), [tests](https://github.com/substack/node-deep-equal/blob/master/test/cmp.js)
- chai's deep-eql [implementation](https://github.com/chaijs/deep-eql/blob/master/lib/eql.js), [tests](https://github.com/chaijs/deep-eql/blob/master/test/eql.js)
- node's deepEqual [implementation](https://github.com/nodejs/node/blob/8c043c12456d9b5a500f9cefcca27a61a1a381cf/lib/assert.js#L145-L182), [tests](https://github.com/nodejs/node/blob/8c043c12456d9b5a500f9cefcca27a61a1a381cf/test/parallel/test-assert.js#L55-L269) (roughly)

Not sure how to go about this, and don't yet know the history behind any of these.

**Comment by jzaefferer** [comment]
For anyone interested in this: Should try to port QUnit's tests to one or all of the other modules and see what fails. We can use those failures to start a discussion with the module owner, along with talking about browser support and whatever else.

**Comment by jzaefferer** [comment]
After talking to @jdalton about this: We should try to replace our `deepEqual` implementation with the one in lodash. We can run our existing tests to see if there are (significant) differences in the design. If it works out well, we could promote the lodash implementation to those other projects listed above.

**Comment by jzaefferer** [comment]
@jdalton I've read https://lodash.com/custom-builds and some other resources, but don't see an option for a (custom) compat build with just isEqual (with dependencies, `global` or `umd` exports). What am I missing?

**Comment by jdalton** [comment]
via lodash-cli:
```bash
lodash include=isEqual
```
of via browserify after you `npm i lodash.isequal`:
```js
browserify -r lodash.isequal -s isEqual -o lodash.isequal.js
```

**Comment by jzaefferer** [comment]
I just tried to replace the `QUnit.equiv` call in `assert.js` to use `_.isEqual` instead, and all tests still pass. That's great!

Still need to figure out how to properly integrate lodash into our build, since this is the first external dependency.

**Comment by jdalton** [comment]
> Still need to figure out how to properly integrate lodash into our build, since this is the first external dependency.

You could go the webpack route and include `isEqual` as either `require('lodash/lang/isEqual')` or `require('lodash.isequal')`.

**Comment by jdalton** [comment]
I can prep a PR tonight to flesh a possible approach out.

**Comment by leobalter** [comment]
That would be great, @jdalton! 

It looks like replacing internal deepEqual is the easiest part, but it needs some good strategy using webpack/browserify.

**Comment by jzaefferer** [comment]
Yeah, switching to internal modules internally is currently the bigger issue. Though if we make that work with webpack, it'll be super easy to add external modules.

**Comment by jzaefferer** [comment]
@jdalton any idea if you'll be able to work on this?

**Comment by jdalton** [comment]
Yes, I'll try to get something working this weekend. Last weekend I was unblocking rollup (es6 module bundler). This weekend can be QUnit bundling time. The switch may want to be put on hold though until after the 4.0 bump as lodash's 3.10.1 doesn't support map/set comparisons.

**Comment by jzaefferer** [comment]
Sounds goood. Making progress on bundling will be valuable either way.

**Comment by jdalton** [comment]
I should revisit this now that v4 is out.

**Comment by jzaefferer** [comment]
That would be great! I think our deepEqual implementation has seen some improvements, but that shouldn't stop us from replacing it.

**Comment by leobalter** [comment]
> I think our deepEqual implementation has seen some improvements, but that shouldn't stop us from replacing it.

yes, I'm looking forward to it.

**Comment by Krinkle** [comment]
Closing for now as I don't expect us to do this any time soon. However if there are one or more persons interested in working on this and have a use case lined up for another library that would consume it, I'd be happy to see this re-open.
